In [135]:
from math import *

**<span style="font-size: 20px; color: blue">(Problem 1) 완전수(perfect number) </span>** 

자연수 $n$에 대하여 $n$이 아닌 모든 약수의 합이 $n$이면 $n$을 완전수라 한다.

예)
- 6의 약수: 1,2,3,6 $\Rightarrow$ $6 = 1+2+3$
- 28의 약수: 1,2,4,7,14,28 $\Rightarrow$ $28 = 1+2+4+7+14$

**1) 자연수 $n$을 입력받아 $n$이 완전수인지 판정하는 is_perfect_number()를 정의하시오.**

예)
- is_perfect_number(6) $\Rightarrow$ True
- is_perfect_number(7) $\Rightarrow$ False

In [136]:
def is_perfect_number(n):
    val = 0
    for i in range(1,n):
        if n%i == 0:
            val += i
    if val == n:
        return True
    else:
        return False

In [137]:
is_perfect_number(6)

True

결과: True

In [138]:
is_perfect_number(10)

False

결과: False

In [139]:
is_perfect_number(28)

True

결과: True

**2) 자연수 $n$을 입력받아 $n$ 이하의 완전수를 모두 출력하는 함수 perfect_number()를 정의하시오.**

In [140]:
def perfect_number(n):
    perfect = []
    for i in range(1,n+1):
        if is_perfect_number(i):
            perfect.append(i)
    return perfect

In [141]:
perfect_number(10000)

[6, 28, 496, 8128]

결과

Perfect number:  [6, 28, 496, 8128]

**<span style="font-size: 20px; color: blue">(Problem 2) 친화수(amicable numbers) </span>** 

두 수의 쌍 $(a,b)$에 대해
- $a$의 약수 중 $a$가 아닌 모든 약수의 합이 $b$이고
- $b$의 약수 중 $b$가 아닌 모든 약수의 합이 $a$이면
두 수 $a$와 $b$를 친화수(또는 우애수)라 한다.

예)
- 220의 약수: 1,2,4,5,10,11,20,22,44,55,110,220 $\Rightarrow$ $284 = 1+2+4+5+10+11+20+22+44+55+110$
- 284의 약수: 1,2,4,71,142,284 $\Rightarrow$ $220 = 1+2+4+71+142$

**1) 자연수 $n$을 입력받아 $n$의 약수 중 $n$이 아닌 모든 약수의 합을 계산하는 divisor_sum()을 정의하시오.**

예)
- divisor_sum(220) $\Rightarrow$ 284

In [142]:
def divisor_sum(n):
    val = 0
    for i in range(1,n):
        if n%i == 0:
            val += i
    return val

In [143]:
divisor_sum(220)

284

결과: 284

**2) 자연수 $n$을 입력받아 친화수 순서쌍 중 작은 수가 $n$ 이하인 친화수 순서쌍을 모두 출력하는 함수 amicable_numbers()를 정의하시오.**

예)
- amicalble_numbers(10000) $\Rightarrow$ [[220, 284], [1184, 1210], [2620, 2924], [5020, 5564], [6232, 6368]]

In [144]:
def amicable_numbers(n):
    friend = []
    for i in range(1,n+1):
        val = divisor_sum(i)
        if  val != i and divisor_sum(val) == i:
            test = 1
            for seq in friend:
                if i in seq:
                    test = 0
            if test:
                friend.append([i,val])
    return friend

In [145]:
amicable_numbers(10000)

[[220, 284], [1184, 1210], [2620, 2924], [5020, 5564], [6232, 6368]]

결과

[[220, 284], [1184, 1210], [2620, 2924], [5020, 5564], [6232, 6368]]

**<span style="font-size: 20px; color: blue">(Problem 3) 우애사슬 </span>** 

자연수의 쌍 $(a_1,a_2,\cdots,a_n)$에 대해
- $a_k$의 약수 중 $a_k$가 아닌 모든 약수의 합이 $a_{k+1}$이고 $(k = 1,\cdots,n-1)$
- $a_n$의 약수 중 $a_n$이 아닌 모든 약수의 합이 $a_1$이면
$(a_1,a_2,\cdots,a_n)$을 **길이 $n$의 우애사슬**이라 한다.

예)
- (12496, 14288, 15472, 14536, 14264): 길이 5 우애사슬

**1) 자연수 $n$과 $M$을 입력받아 최소한 한 원소가 $M$ 이하인 길이 $n$ 우애사슬을 모두 출력하는 함수 chain()을 정의하시오**

예)
- chain(1, 1000) $\Rightarrow$ [[6], [28], [496]]
- chain(2, 10000)  $\Rightarrow$ [[220,284], [1184,1210], [2620,2924], [5020,5564], [6232, 6368]]
- chain(5, 20000)  $\Rightarrow$ [[12496, 14288, 15472, 14536, 14264]]

In [146]:
def chain(n,M):
    result = list()
    for i in range(1,M+1):
        sequence = [i]
        tmp = i
        for j in range(1,n):
            val = divisor_sum(tmp)
            if val not in sequence:
                sequence.append(val)
                tmp = val
            else:
                break
        if len(sequence) == n and divisor_sum(sequence[-1]) == sequence[0]:
            test = 1
            for seq in result:
                if sequence[0] in seq:
                    test = 0
            if test:
                result.append(sequence)
    return result

In [147]:
chain(2, 10000)

[[220, 284], [1184, 1210], [2620, 2924], [5020, 5564], [6232, 6368]]

결과:
    
[[220, 284], [1184, 1210], [2620, 2924], [5020, 5564], [6232, 6368]]

In [148]:
chain(5, 20000)

[[12496, 14288, 15472, 14536, 14264]]

결과:

[[12496, 14288, 15472, 14536, 14264]]